### DISCUSSION TITLE
Strings: Add option to parse placeholders from the key instead of the value.

### DISCUSSION THREAD

**Discussion by CraigSiemens** [SwiftGen/SwiftGen#900] [Ideas]
We're looking into adding codegen for out strings files. In our codebase we've got the placeholders defined in the keys like this.
```
/* Comment */
"key-$@-%d" = "Something %d something %@";
```

The reason we have this is because we'd used `genstrings` to create our strings files which automatically added the positions to the placeholders in the value field if needed like `%1$@, %2$@, etc`.

This causes two issues for us when using `SwiftGen`.
1. The generated method name for the above example would be `keyD` since it doesn't remove placeholders before creating the name.
1. We never change the order of the placeholders in the key but frequently change them in the value due to changing design requirements. If we change the order of placeholders in the value, we'll have to make code changes to support it which we'd like to avoid.

Is this something SwiftGen would be interested in supporting? I can think of two possible ways it could be added.

1. Try parsing placeholders from the key first, if none found, try parsing from the value.
1. Add an option for whether to parse placeholders from the key or value, default to value if backwards compatibility is a concern.

**Comment by djbe** [comment]
I'll take a deeper look at this tomorrow, but as a quick FYI, the **order** of the placeholders in the value are very important, not only for SwiftGen but also for `String(format:...)`.

SwiftGen will correctly parse the placeholders so long as you add the position to them (like you mentioned), such as `%1$@`. It's very important that these positions, or this order, is consistent between translations, otherwise you'll get unexpected results and/or crashes.

As for the key names, as a quick suggestion you could use a custom template to ignore parts of the key when generating Swift code.

**Comment by CraigSiemens** [comment]
Another reason we like having the placeholders in the keys is that we have our unit tests set to not use the translated values. That means any string formatting will use the key instead of the value which means we don't have to update our tests if the translation changes. Then we can compare against `key-SomeString-123` to ensure the output is using the right format and arguments.

**Comment by CraigSiemens** [comment]
While setting up the strings SwiftGen in our app, I realized there could be another benefit from parsing the placeholders from the key, being able to add labels to the parameters in the generated methods.

## Current behaviour

For example, the following string file
```
"bananas.owner" = "Those %d bananas belong to %@.";
```
would currently generate the methods
```
L10n.Bananas.owner(_:_:)
```
The more parameters that are added, the less clear it becomes which position that would co in, especially if they're all the same type.

## Possible new behaviour

Adding the placeholders to the key in the simplest way like `bananas.owner-%d-%@` would still generate the same method, but if words were placed between the placeholders like the following.
```
"bananas.owner-%@-count-%d" = "Those %d bananas belong to %@.";
```
The code generation could use those words between the placeholders as possible labels.
```
// Current codegen
L10n.Bananas.ownerCount(_:_:)

// Keep the "structured" behaviour and make a function for everything after "."
L10n.Bananas.owner(_:count:)

// A special version of "structured" where the last `keyStructure` is used to label the parameters.
L10n.bananas(owner:count:)
```
